langchain/logging/sqlite.py

```diff
@@ -0,0 +1,354 @@
+                                        )
+                                    ]
+                                )
+                            ]
+                        )
+                    ]
+                )
+            ]
+        )
+        session.add(zeroshot_run)
+        session.commit()
+
+    from sqlalchemy.orm import joinedload
+    from sqlalchemy import select
+
+    stmt = select(ChainRun).where(ChainRun.id == 1).options(joinedload(ChainRun.child_llm_runs),
+                                                            joinedload(ChainRun.child_chain_runs).joinedload(
+                                                                ChainRun.child_llm_runs),
+                                                            joinedload(ChainRun.child_tool_runs))
+    zeroshot_chain = session.scalars(stmt).unique().one()
+    print_run(zeroshot_chain, "")
+
+
+def print_run(run: Union[LLMRun, ChainRun, ToolRun], tabs: str) -> None:
+    print(f"{tabs}Run: {run}")
+    if isinstance(run, ChainRun) or isinstance(run, ToolRun):
+        all_runs = run.child_chain_runs + run.child_llm_runs + run.child_tool_runs
+        all_runs.sort(key=lambda x: x.execution_order)
+        for child_run in all_runs:
+            print_run(child_run, tabs + "\t")
+
+
+if __name__ == "__main__":
+    main()
+
+
+class LoggerException(Exception):
+    """Base class for exceptions in logging module."""
+
+
+class SqliteLogger(BaseLogger):
+    """A logger that stores the logs in a sqlite database."""
+
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(SqliteLogger, cls).__new__(cls)
+            # TODO: make the db initialization string an environment variable
+            cls._instance._db = create_engine("sqlite://", echo=False, future=True)
+            Base.metadata.create_all(cls._instance._db)
+            cls._instance._stack = []
+            cls._instance._session = Session(cls._instance._db)
+            cls._instance._execution_order = 1
+        return cls._instance
+
+    def _log_run_start(self, run: Union[LLMRun, ChainRun, ToolRun]) -> None:
+        """Log the start of a run."""
+
+        if self._stack:
+            if not (isinstance(self._stack[-1], ChainRun) or isinstance(self._stack[-1], ToolRun)):
+                self._session.rollback()
+                raise LoggerException(f"Nested {run.__class__.__name__} can only be logged inside a ChainRun or ToolRun")
+            if isinstance(run, LLMRun):
+                self._stack[-1].child_llm_runs.append(run)
+            elif isinstance(run, ChainRun):
+                self._stack[-1].child_chain_runs.append(run)
+            else:
+                self._stack[-1].child_tool_runs.append(run)
+        self._stack.append(run)
+        self._session.add(run)
+
+    def _end_log_run(self) -> None:
+        """Call at the end of a run."""
+
+        self._execution_order += 1
```

Comment:
so execution order is based on the ending? wouldnt it be more intuitive to have it based on the start?

Reply:
I think either one works

Reply:
Actually, execution order in this convention is helpful for pretty printing in the correct order, so I'd like to keep it this way